crates/sui-indexer-alt/src/handlers/obj_info_v2.rs

```diff
@@ -0,0 +1,1724 @@
+impl Handler for ObjInfoV2 {
+    type Store = Db;
+
+    const PRUNING_REQUIRES_PROCESSED_VALUES: bool = true;
+
+    async fn commit<'a>(values: &[Self::Value], conn: &mut Connection<'a>) -> Result<usize> {
+        let stored = values
+            .iter()
+            .map(|v| v.try_into())
+            .collect::<Result<Vec<StoredObjInfoV2>>>()?;
+        Ok(diesel::insert_into(obj_info_v2::table)
+            .values(stored)
+            .on_conflict_do_nothing()
+            .execute(conn)
+            .await?)
+    }
+
+    async fn prune<'a>(
+        &self,
+        from: u64,
+        to_exclusive: u64,
+        conn: &mut Connection<'a>,
+    ) -> Result<usize> {
+        let query = postgres::sql_query!(
+            "
+            WITH predecessors AS (
+                SELECT
+                    latest.object_id as latest_object_id,
+                    latest.cp_sequence_number as latest_cp_sequence_number,
+                    latest.marked_predecessor as latest_marked_predecessor,
+                    pred.object_id as pred_object_id,
+                    pred.cp_sequence_number as pred_cp_sequence_number,
+                    pred.marked_predecessor as pred_marked_predecessor
+                FROM obj_info_v2 latest
+                LEFT JOIN LATERAL (
+                    SELECT object_id, cp_sequence_number, marked_predecessor
+                    FROM obj_info_v2 p
+                    WHERE p.object_id = latest.object_id
+                    AND p.cp_sequence_number < latest.cp_sequence_number
+                    ORDER BY p.cp_sequence_number DESC
+                    LIMIT 1
+                ) pred ON true
+                WHERE latest.cp_sequence_number >= {BigInt} AND latest.cp_sequence_number < {BigInt}
+                ORDER BY latest.object_id, latest.cp_sequence_number
+                FOR UPDATE OF latest
+            ),
+            -- Otherwise, flag them for later deletion
+            pred_obsolete AS (
+                UPDATE obj_info_v2
+                SET obsolete_at = p.latest_cp_sequence_number
+                FROM predecessors p
+                WHERE obj_info_v2.object_id = p.pred_object_id
+                AND obj_info_v2.cp_sequence_number = p.pred_cp_sequence_number
+                AND p.pred_marked_predecessor = false
+                RETURNING obj_info_v2.object_id
+            ),
+            -- Finally, mark 'latest' rows to have marked their own immediate predecessors
+            marked_predecessor AS (
+                UPDATE obj_info_v2
+                SET marked_predecessor = true
+                WHERE (object_id, cp_sequence_number) IN (
+                    SELECT latest_object_id, latest_cp_sequence_number FROM predecessors
+                    ORDER BY latest_object_id, latest_cp_sequence_number
+                )
+                RETURNING object_id
+            ),
+            -- Delete preds that already marked their own immediate predecessors
+            pred_deleted AS (
```

Comment:
To make sure I understand, `pred_deleted` and `intermediate_deleted` are optional here, meaning that even if they are not executed, these entries will be deleted in T1 anyway. Is that right? We are just conveniently deleting them in T0 because they have already been loaded.